### PR TITLE
rewrite synchronization phases to isolate individual operations and allow for graceful failure continuation of a run

### PR DESCRIPTION
**Pull request by k-rister**

### CODE REVIEW

**Comment by k-rister** [comment]
This PR is currently blocked from passing CI while waiting for a bench-uperf PR to be merged (https://github.com/perftool-incubator/bench-uperf/pull/39).

**Comment by k-rister** [comment]
I'm looking to start getting some feedback on this PR.  I've decided to defer switching over to the [new roadblock code](https://github.com/perftool-incubator/roadblock/pull/36) as a future change.  For one, the roadblock code hasn't even been merged yet but more importantly, I think there is substantial potential for improvements from just these changes alone to justify getting them into the hands of users sooner rather than later.

I'm hoping to hear from @mrbojangles3 about testing uperf on a real K8S cluster to ensure that I haven't broken the message passing that is used for setting up the various different modes that are supported

**Review by rafaelfolco — DISMISSED** [review]
honestly I got lost while reviewing this giant PR. So my suggestion is to merge if you are confident and test it. Revert if it breaks something.

File: engine/bootstrap
Comment (k-rister) [reply engine/bootstrap:233]: The `/usr/local/bin/client-server-script-library` file is just a bunch of functions that used to exist in `/usr/local/bin/client-server-script` that have been pulled out and put into a library file.  The intention of this is that the library file can continue to be used by `/usr/local/bin/client-server-script` while also providing a mechanism to call those functions from other scripts and/or processes.  This part of the PR is a precursor to the adopting the pending roadblock changes so its somewhat a noop at this time given that I deferred that work until a future PR.  As a reminder, the roadblock changes will allow a roadblock invocation to block on a subprocess, waiting for it to complete before finishing the roadblock.  Some of the actions that we want to invoke as a subprocess of roadblock are what has been moved from `client-server-script` to `client-server-script-library` to eventually facilitate that.

I did consider deferring the creation of `client-server-script-library` until that future PR, but ultimately decided that I had made significant changes to the contents of `client-server-script-library` as part of this PR that would have been to mapped back to `client-server-script` prior to the change, but I think that would be a significant amount of rework for questionable gain.  In this way, deferring the adoption of the roadblock changes has made this PR a bit confusing :/

Ultimately, to answer your question `client-server-script` is still the script we are calling, but it depends heavily on `client-server-script-library` so I'm checking for the presence of both before execing the one.

File: endpoints/base
Comment (k-rister) [reply endpoints/base:417]: Sort of.  I determined that these were blocks of code that were only used by the localhost endpoint and would have been orphaned if they had left behind.

It is important to remember that @atheurer wrote the localhost endpoint as a means of doing the initial bootstrap of rickshaw and that is has been completely idle and abandoned ever since.  As such, all of that code has been completely unmaintained for awhile and left to rot...these functions are a perfect example of something that is no longer being called by the actively used and maintained endpoints.

File: endpoints/base
Comment (k-rister) [reply endpoints/base:742]: So one of the things I tried to do here is to actually unroll the loops that were present.  The way it was originally written the loops would iterate over a set of synchronization points (ex. `A`, `B`, & `C`) and create `A-start`, `B-start`, and `C-start` while also creating (in reverse order) `C-stop`, `B-stop`, and `A-stop`.  The problem with this was when you got a roadblock failure during a run, lets say at `B-stop`, you would go look at the code and there was no way to easily find where the `B-stop` synchronization point was because it didn't actually exist as that name anywhere in the code -- the name was pieced together as the loop was iterated over and various conditional statements were evaluated.  You would eventually be able to locate the problem if you knew the code really well and could follow all the logic but it was a very high barrier to entry.

To make it even more complicated this was happening in 3 different scripts (`rickshaw-run`, `endpoints/base`, and `engine/client-server-script`) while being implemented in two different languages (`rickshaw-run` is in perl, `endpoints/base` and `engine/client-server-script` are in bash shell scripts) and the conditionals were different between all 3 scripts because each one is doing slightly different things depending on it's role in the current synchronization point.

The big change here, other than a renaming of the synchronization points and the creation of some new ones is that that loop logic has been eliminated.  This allows a user to go to one of the scripts with the name of a synchronization point they failed at and easily search for and find that point in the code.  They can then see exactly what is happening at that synchronization point and try to pinpoint what the problem is.

The other thing that was happening in the loops is there was a lot of complex logic being used to track progress so that it would know what had already happened so that it would know what to do in the case of a failure -- ie. since `A-start` happened we need to create a `A-stop`.  This dynamic behavior resulted in us not doing proper cleanup and continuation in the face of errors which is one of the reasons why a failed run would exit uncleanly.  The new code tries to continue all the way to completion as much as possible once a certain point is reached -- namely the test phase.  The tracking that was being done doesn't need to be done if we are always trying to run to completion by processing all synchronization points every time, it just needs to know if there is some type of error (ie. an abort or a quit) and how to properly handle that (ie. skip running tests when a quit is signaled but continue making progress).  The other thing that happens is that if a timeout is observed, the only way to successfully make progress through the remaining roadblocks is to remove the follower(s) that did not check in to a roadblock from the future roadblocks so that they do not continue to cause timeouts.

You can probably sense a theme of trying to make it easier to follow the code for trouble shooting reasons, so I also tried to make the code as similar as possible across the 3 different lines of execution.  I changed some of the function names so that they are the same and tried to create commonly named and functional new ones as much as possible.  In some cases the implementation details are slightly different but the overall intent is supposed to be the same and therefore make it easier to follow -- at least that is the hope/intent.  I've also tried to move chunks of the logic that were formerly in the loops into functions so that it can easily be called without negatively impacting the ability to follow the flow of exection.

You are accurate in saying that this is very hard to follow due to it's size, but the unfortunate reality is that since we have these 3 different scripts that synchronize with each other as they run and it was originally done with the highly integrated loops that it is very difficult, if not impossible, to break up the changes into smaller more incremental pieces.  I'm hoping that going forward that is not the case from now on.  For example, it should be relatively straight forward to insert one or more new synchronization points at any point in the process now across all 3 scripts with minimal work, I think.

**Comment by k-rister** [comment]
@rafaelfolco Bottom line, I'm hopeful that the new code should actually be fairly easy to understand and follow.  The problem is that the old code is not and so understanding the changes is very, very difficult.

**Comment by k-rister** [comment]
@rafaelfolco That said, I would appreciate just looking over the new code for things I obviously did wrong such as typos or basic logic errors.  With the exception of the K8S endpoint and specifically it's uperf work I'm fairly confident in this code at this point.  I have run it against oslat (remotehost), fio (remotehost), uperf (microk8s), and trafficgen (remotehost, no server) at home and both fio and uperf (remotehost and microk8s) in CI and it appears to be working in all of those scenarios so far.

**Comment by k-rister** [comment]
I think this may be finally done...but will probably wait until next week to merge.  Feel free to review and/or make comments in the interim.

**Pull request merged**
